src/sql/workbench/contrib/notebook/browser/markdownToolbarActions.ts

```diff
@@ -87,7 +92,11 @@ export class MarkdownTextTransformer {
 					if (isUndo) {
 						this.handleUndoOperation(markdownLineType, startRange, endRange, editorModel, beginInsertedCode, endInsertedCode, selections, selection);
 					} else {
-						this.handleTransformOperation(markdownLineType, startRange, endRange, editorModel, beginInsertedCode, endInsertedCode, selections, selection);
+						if (markdownLineType === MarkdownLineType.BEGIN_LINE) {
+							this.handleHeadingTransformOperation(singleLineStartEndRange, editorModel, beginInsertedCode);
```

Comment:
This successfully adds a user-selected heading level to a line that doesn't already have a heading level.

What I need next is code that checks for existing heading and replaces if with the level selected by the user, or remove it if the user selected Paragraph rather than a heading level.

Reply:
So I don't think that you should have a special "heading" transform option. There's basically only 3 things you could ever do to a line : 

1. Undo (remove)
2. Replace
3. Add

Right now we have code for 1 & 3, but you're going to want to add support for #2 as well. 

So the cases you care about are : 

1. Adding a heading when none exists (#3 - this you already have)
2. Removing a heading when that heading already exists on that line OR the MarkdownButtonType is paragraph . This should be handled by the isUndo operation - the "heading already exists" should be handled by the existing logic. You'll want to modify it to special case the paragraph though
3. Replacing  - this will be a new "operation" you add. So I'd add another check isReplaceOperation very similar to the isUndoOperation which does the checks as needed (I'd probably add a mapping of "related" types that should replace each other to keep it generic since the logic is going to be the same except for the specific strings you're checking for)

Reply:
@Charles-Gagnon I modified transformText to help with undo operation of a single line (with only cursor position) addition of heading or list item. Undo works with multi-line addition as well.

I need help with the Replacing of headings or list items in the case when the user selects Paragraph, a different heading level or list item.